### ISSUE TITLE
`mod_blocking` doesn't stop messages from being processed by `modules.mod_mam.pm` or `mod_inbox`

### ISSUE THREAD

**Issue by DoxasticFox** [esl/MongooseIM#4119]
I'm using the docker image `mongooseim/mongooseim:6.1.0-5-gabdcd0b48`.

Steps to reproduce the issue:

1. Enable `mod_blocking`, `modules.mod_mam.pm` and `mod_inbox`
1. Create two users, Alice and Bob. Let them have active sessions at the same time
1. Let Alice block Bob.
1. Let Alice attempt to message Bob.
1. Let Bob also attempt to message Alice.
1. Neither Bob nor Alice will see each other's messages---So far, so good.
1. But now let Bob and Alice start new sessions then query MAM or their respective inboxes with the typical stanzas:
   ```xml
   <iq type='set' id='${queryId}'>
     <query xmlns='urn:xmpp:mam:2' queryid='${queryId}'>
       <x xmlns='jabber:x:data' type='submit'>
         <field var='FORM_TYPE'>
           <value>urn:xmpp:mam:2</value>
         </field>
         <field var='with'>
           <value>${jid}</value>
         </field>
       </x>
       <set xmlns='http://jabber.org/protocol/rsm'>
         <max>50</max>
         <before/>
       </set>
     </query>
   </iq>
   ```

   ```xml
   <iq type='set' id='${queryId}'>
     <inbox xmlns='erlang-solutions.com:xmpp:inbox:0' queryid='${queryId}'>
       <x xmlns='jabber:x:data' type='form'>
         <field type='text-single' var='box'><value>${box}</value></field>
       </x>
     </inbox>
   </iq>
   ```

   Unexpectedly, some messages sent while blocked will appear! Specifically, messages sent from Bob to Alice will be present in MAM and her Inbox! But, correctly, none of the messages which Alice sent will be displayed in MAM or either party's inbox.

**Comment by arcusfelis** [comment]
Not a bug, just a pretty **inconsistent** XMPP behaviour.
mod_blocking is based on mod_privacy, and mod_privacy needs some list to be activated on connect (but I think there is a default list too).

mod_mam has its own "privacy" settings. i.e. before they were in XEP-0313, now they got a separate XEP:
Oh, there is literally this XEP too https://xmpp.org/extensions/attic/xep-0441-0.2.0.html

Preferences ARE supported, you can do it.

```xml
<iq type='set' id='juliet3'>
  <prefs xmlns='urn:xmpp:mam:2' default='roster'>
    <always>
      <jid>romeo@montague.lit</jid>
    </always>
    <never>
      <jid>montague@montague.lit</jid>
    </never>
  </prefs>
</iq>
```

I.e. you need to set list of jids to block once for mod_blocking and once for MAM preferences.

Theoretically we could just copy mod_mam_rdbms_prefs module to create mod_mam_blocking_prefs and hook it to use the list from mod_blocking. Except, it makes prefs read only (i.e. only updatable though mod_blocking API).

Except, it would not fully fix mod_inbox blocking.
Except, the behaviour for multichat rooms is undefined (i.e. you probably wanna put a spoiler on the client for that and adjust the push settings, but it would require work on the client side and someone actually describing what is a preferred behaviour).

Also, maybe we could still want to archive blocked threads (in case we want to unblock the user later), but don't show such threads. But in this case these messages should be filtered out from the list of messages we query when we don't provide `with` filter (maybe we should have a filter to filter out blocked messages, but good luck to make that efficient on the backend).

At least push service respects mod_blocking (I think).

TLDR: we could integrate MAM with mod_blocking, it would not fix mod_inbox. + the whole behaviour is still not consistent or desired. Once we figure out the consistent behaviour or better XMPP people would make a XEP describing a preferred solution - we could implement it.
